### PR TITLE
feat(v6): cross-platform parity audit remediation (interceptor, events, attributes, missing APIs)

### PR DESCRIPTION
**Pull request by kherembourg**
## Context

Cross-platform parity audit of the Flutter SDK bridge against the pinned native iOS/Android SDKs (base branch `feat/sdk-v6-migration`). The audit found 12 confirmed divergences (crashes, silent field loss, missing handlers, dead code) plus a set of missing APIs. This PR remediates all of them, adds the missing APIs, and ports/updates the parity test suite. Every fix keeps the wire contract identical across iOS and Android.

## Critical fix

**FLT-W-08 / REC-01 — web_checkout action interceptor was broken end-to-end on iOS.**
iOS serialized `webCheckoutProvider` as the native enum's raw `Int` (`PLYWebCheckoutProvider`: Stripe=0/Other=1/None=2), but Dart casts that field `as String?`. A non-null int triggered a `_CastError` *outside* the interceptor's try/catch, so: the app's `interceptAction(webCheckout, …)` handler never fired, and the native SDK's pending completion callback was never resolved (SDK left hanging). iOS now sends the case-name string (`"STRIPE"` / `"OTHER"`, key omitted for the `.none` sentinel), exactly matching Android's `action.webCheckoutProvider.name`. Verified the native strings match: native Android `PLYWebCheckoutProvider` is `STRIPE`/`OTHER`, iOS maps `.stripe→"STRIPE"`, `.other→"OTHER"`. Dart also gained a tolerant parser accepting both `String` and the legacy `Int` (belt-and-suspenders), with tests for both formats.

## Fixes by ID

- **FLT-W-01 / REC-04** — Android had no `signPromotionalOffer` case → `MissingPluginException`. Now resolves success with an empty map (StoreKit-only feature, no Google Play equivalent); Dart doc marks it iOS-only and never-rejecting.
- **REC-09 / FLT-W-04** — `mapType()` gained the real iOS `DICTIONARY` case and now maps any unknown type to `PLYUserAttributeType.unknown` (+ log) instead of throwing an uncaught `ArgumentError` inside the `.listen()` callback.
- **REC-09 / FLT-W-07** — `subscriptionSource` null / out-of-range (Android emits `null` for a store type outside the 4 known ones) now degrades to `PLYSubscriptionSource.none` instead of crashing the whole `userSubscriptions()` future on a `List` index error.
- **FLT-W-06 / REC-09** — iOS `presentation.contentId` was hardcoded `null`. The native `PLYPresentation` exposes no `contentId` getter, so the plugin now keeps a per-`requestId` registry of the value passed to the request builder and echoes it back in `presentationToMap`. (This is the real contentId from the builder, not a placeholder.)
- **REC-13 / EVT-01** — Unknown event names silently became `APP_CONFIGURED`. Added a `PLYEventName.UNKNOWN` sentinel + shared `_eventNameFromWire` fallback, plus the two missing native events `PLACEMENT_OPENED` and `PURCHASE_FROM_STORE_TAPPED`. The stale "46 cases" comment is corrected to the accurate count (51 = 50 native + UNKNOWN); matching is by exact case name, never ordinal.
- **FLT-W-05** — iOS `setLanguage` never called `result(...)` (would hang an awaited call). Now calls `result(true)`.
- **FLT-W-10** — `runningMode` Int-parsing branches on both natives were dead *and* wrong (native iOS rawValues are Observer=2/Full=3, not 0/1). Removed; only the string contract (`"observer"`/`"full"`, fallback observer) remains. `full` still verified end-to-end.
- **PAR-27** — `PLYLogLevel` is now wire-encoded as `.name` everywhere on the Dart side (both `setLogLevel` and the `start()` builder), parsed by the same tolerant parser on both natives, instead of an Int index that would `ClassCastException`/silently default.
- **FLT-W-12** — Inline `NativeView` dismiss fallback no longer registers the global `setEventDelegate` (which the codebase itself flags as less reliable in v6 RC+ and which had no unset). It uses the closure-based `setEventCallback`, reset to a no-op on `deinit`.
- **FLT-W-03 / PAR-11** — Android `removeDefaultPresentationDismissHandler` now overwrites the native handler with a no-op at the native level (`setDefaultPresentationDismissHandler { }`), instead of only nulling the Dart-side reference while the native listener kept firing.
- **PAR-13** — `PurchaselyBridge` (`ensureInstalled`/`debugReset`) is a test-only entry point; removed from the public barrel export. Tests import `src/bridge.dart` directly.

## New APIs

- `Purchasely.closeAllScreens()` (PAR-19) → both natives.
- `Purchasely.getBuiltInAttributes()` / `getBuiltInAttribute(key)` (PAR-07) → both natives.
- `userSubscriptions({invalidateCache})` / `userSubscriptionsHistory({invalidateCache})` (PAR-29) → both natives. **See behavior note below.**
- `userLogout({clearUserAttributes = true})` (PAR-30) → both natives.
- `addEventListener` / `removeEventListener` aliases for `listenToEvents` / `stopListeningToEvents` (REC-18).
- `PLYAttribute.oneSignalUserId` (REC-11 / ENM-03) added to all three synchronized enums (Dart, iOS `FlutterPLYAttribute`, Android `FlutterPLYAttribute`) in identical order; native maps by symbol. No `oneSignalPlayerId` in the Dart enum (iOS-only native case; the three bridges map by ordinal so it must stay out).

## Behavior notes

- **PAR-29 changes Android's default.** The Android plugin previously hardcoded `userSubscriptions(true)` / `userSubscriptionsHistory(true)` (always invalidate cache). It now defaults `invalidateCache` to **false** to match the spec and iOS (whose no-arg call already defaulted to false). Net effect: on Android, `userSubscriptions()` / `userSubscriptionsHistory()` with no argument now returns cached results by default instead of forcing a refresh. Callers who need the old behavior pass `invalidateCache: true`. iOS behavior is unchanged.
- **Events**: an unrecognized native event name now maps to `PLYEventName.UNKNOWN` (logged), never silently to `APP_CONFIGURED`.
- **PAR-23 — `PLYPresentation.close()` on Android**: the call is routed through the per-presentation handle (`loadedPresentations[requestId]?.close()`), mirroring iOS's structure. However, in the pinned native Android SDK, `PLYPresentationBase.Loaded.close()` currently delegates to `Purchasely.closeAllScreens()` — so **today the Android call closes ALL screens regardless of `requestId`** (this is pinned by a native-side test: `Loaded close delegates to Purchasely closeAllScreens without throwing`). The handle routing is kept because it becomes scoped transparently, with no Flutter-side change, once the native SDK implements per-presentation close. iOS already scopes to the tracked `requestId` (falling back to closing everything only when no handle matches).
- **`signPromotionalOffer` on Android**: no-op that resolves success with an empty map (never rejects) — meaningful only on iOS.

## rc.4 hardening

`plyPlanTypeFromWire` tolerates both the `String` wire form (`RENEWING_SUBSCRIPTION` → autoRenewing, etc.) and the legacy `Int` index, degrading unknown strings to `PLYPlanType.unknown` rather than throwing. (The native Android SDK has emitted the string form since 2021; the Int tolerance is a safety belt.) Test added.

## Known issue — out of scope

The **iOS example build is currently broken on the base branch** `feat/sdk-v6-migration` by 3 pre-existing symbol mismatches (`PLYDisplayMode` vs `PLYTransition`, `builder.onClose` vs `onCloseRequested`, `.id` vs `.screenId`). These are fixed separately in Kevin's branch `fix/ios-plytransition-type` and are intentionally **not** touched here (touching them would collide with that branch). Confirmed these 3 lines are identical on the base branch and untouched by this PR's diff. This PR's iOS changes were compile-verified against the vendored `Purchasely-Swift.h` (all native symbols used exist: `getBuiltInAttributes`, `getBuiltInAttributeWith:`, `setEventCallback`, `userLogout:(BOOL)`, `userSubscriptions:(BOOL)…`, `closeAllScreens`, `PLYAttributeOneSignalUserId`); a full iOS build additionally requires the 3 base lines patched locally.

## Tests (actually run)

- **Dart** `flutter test`: **293 passed, 0 failures**.
- `flutter analyze`: **No issues found**.
- `dart format --set-exit-if-changed .`: **clean (0 changed)**.
- **Android native** `:purchasely_flutter:testDebugUnitTest` (via `flutter build apk --debug` then Gradle, `--rerun-tasks`): **12 tests, 0 failures, 0 errors**. Example APK builds successfully.
- **iOS native**: not runnable — base branch iOS example build is broken by the 3 pre-existing bugs above (out of scope). iOS changes compile-verified against the vendored SDK header.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: purchasely/ios/Classes/NativeView.swift [hunk 1]
```diff
@@ -20,7 +20,21 @@ class NativeView: NSObject, FlutterPlatformView {
...
+        // `SwiftEventHandler` prefers over the ObjC delegate API
+        // (`setEventDelegate`/`PLYEventDelegate`), which the SDK's own comment
+        // there flags as less reliable in v6 RC+ (FLT-W-12). `setEventCallback`
+        // is a single global slot with no "add" variant, so this is reset to a
+        // no-op on `deinit` for a clean handoff instead of leaving a stale
+        // closure referencing a deallocated view.
+        Purchasely.setEventCallback { [weak self] event, _ in
+            guard event == .presentationClosed else { return }
+            self?.handlePresentationClosed()
+        }
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/ios/Classes/NativeView.swift]: Fixed in b2f843d — NativeView now uses `Purchasely.setEventDelegate(self)` (the independent PLYEventDelegate slot) and `removeEventDelegate()` in `deinit`, so it no longer clobbers SwiftEventHandler's `setEventCallback` stream. Added an E2E guard (`inline_events_test.dart`, wired into both runners) asserting a global `listenToEvents` listener keeps receiving paywall events while the inline view is mounted.

File: purchasely/lib/purchasely_flutter.dart [hunk 2]
```diff
@@ -585,6 +640,38 @@ class Purchasely {
...
+        .invokeMethod('getBuiltInAttribute', <String, dynamic>{'key': key});
+
+    try {
+      value = DateTime.parse(value);
+    } catch (FormatException) {
+      //do nothing it is not a date
+    }
+
+    return value;
+  }
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/lib/purchasely_flutter.dart:663]: Fixed in b2f843d — replaced `catch (FormatException)` (which binds a variable named FormatException and catches everything) with an explicit `catch (_)` + comment. The broad catch is intentional: a non-String value throws TypeError from DateTime.parse and must be swallowed. Pinned by a new Dart test (`userAttribute returns a non-string value as-is`).

File: purchasely/lib/purchasely_flutter.dart [hunk 3]
```diff
@@ -585,6 +640,38 @@ class Purchasely {
...
+
+    return attributes.map((key, value) {
+      dynamic attributeValue = value;
+      try {
+        attributeValue = DateTime.parse(value);
+      } catch (FormatException) {
+        //do nothing it is not a date
+      }
+      return MapEntry(key, attributeValue);
+    });
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/lib/purchasely_flutter.dart:680]: Fixed in b2f843d — the same `catch (_)` change was applied to all four sites (userAttribute/userAttributes/getBuiltInAttribute/getBuiltInAttributes).

File: purchasely/android/src/main/kotlin/io/purchasely/purchasely_flutter/PurchaselyFlutterPlugin.kt [hunk 4]
```diff
@@ -409,6 +428,11 @@ class PurchaselyFlutterPlugin: FlutterPlugin, MethodCallHandler, ActivityAware,
                 clearBuiltInAttributes()
                 result.safeSuccess(true)
             }
+            "getBuiltInAttributes" -> getBuiltInAttributes(result)
+            "getBuiltInAttribute" -> {
+                val key = call.argument<String>("key") ?: return
+                getBuiltInAttribute(key, result)
+            }
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/android/src/main/kotlin/io/purchasely/purchasely_flutter/PurchaselyFlutterPlugin.kt:439]: Fixed in acfc846 — a missing `key` now completes with `result.error("MISSING_PARAMETER", …)` instead of a bare `return`, so the awaited Dart Future no longer hangs. Matches the setDebugMode validation pattern; added an Android unit test (`getBuiltInAttribute without key returns argument error instead of hanging`).

File: purchasely/ios/Classes/NativeView.swift [hunk 5]
```diff
@@ -20,7 +20,21 @@ class NativeView: NSObject, FlutterPlatformView {
...
+        // below) doesn't fire for the embedded controller, synthesise the
+        // dismissal from the `.presentationClosed` SDK event instead. Uses the
+        // closure-based `setEventCallback` API — the same one
+        // `SwiftEventHandler` prefers over the ObjC delegate API
+        // (`setEventDelegate`/`PLYEventDelegate`), which the SDK's own comment
+        // there flags as less reliable in v6 RC+ (FLT-W-12). `setEventCallback`
+        // is a single global slot with no "add" variant, so this is reset to a
+        // no-op on `deinit` for a clean handoff instead of leaving a stale
+        // closure referencing a deallocated view.
+        Purchasely.setEventCallback { [weak self] event, _ in
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/ios/Classes/NativeView.swift]: Fixed in b2f843d — NativeView switched to `setEventDelegate` (a separate slot from `setEventCallback`), so it no longer takes over SwiftEventHandler's global callback.

File: purchasely/ios/Classes/NativeView.swift [hunk 6]
```diff
@@ -127,9 +141,28 @@ class NativeView: NSObject, FlutterPlatformView {
...
+    }
+
     deinit {
         NotificationCenter.default.removeObserver(self)
         UIDevice.current.endGeneratingDeviceOrientationNotifications()
+        // Clean unregistration (FLT-W-12): release the closure's `self`
+        // capture instead of leaving a stale callback referencing a
+        // deallocated view. `setEventCallback` has no "remove" counterpart —
+        // overwriting with a no-op is the cleanest handoff the API allows.
+        Purchasely.setEventCallback { _, _ in }
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/ios/Classes/NativeView.swift]: Fixed in b2f843d — `deinit` now calls `Purchasely.removeEventDelegate()`, which releases only the delegate slot and leaves SwiftEventHandler's `setEventCallback` stream intact.

File: purchasely/ios/Classes/SwiftPurchaselyFlutterPlugin.swift [hunk 7]
```diff
@@ -12,6 +12,12 @@ public class SwiftPurchaselyFlutterPlugin: NSObject, FlutterPlugin {
     // Dart Presentation handle; there is no native dispose API yet.
     static var requests: [String: PLYPresentationRequest] = [:]
     static var loadedPresentations: [String: PLYPresentation] = [:]
+    // The native `PLYPresentation` protocol and `PLYPresentationRequest`
+    // protocol expose no `contentId` getter (verified against the SDK
+    // interface) — only the builder's write-only `.contentId(_:)` setter. Keep
+    // our own registry of what was passed in so `presentationToMap` can echo
+    // it back instead of hardcoding null (FLT-W-06 / REC-09).
+    static var requestContentIds: [String: String] = [:]
```
[the remark this answers is not in the document]
Comment (kherembourg) [reply purchasely/ios/Classes/SwiftPurchaselyFlutterPlugin.swift:20]: Fixed in b2f843d — `requestContentIds.removeValue(forKey:)` is now called on every dismiss path: NativeView.emitDismissed (inline), builder.onDismissed (full-screen), and the display-error branch.

**Comment by kherembourg** [comment]
### Review findings addressed

All 7 inline findings (Greptile + Copilot) resolved, plus added test coverage for the inline event path and the Dart/Android edge cases.

| # | Finding | Fix | Commit |
|---|---------|-----|--------|
| 1 | `NativeView` `setEventCallback` clobbers `SwiftEventHandler`'s global event stream | Use the independent `PLYEventDelegate` slot (`setEventDelegate`/`removeEventDelegate`) | `b2f843d` |
| 2 | `deinit` resets the global callback to a no-op → events never recover | `deinit` releases only the delegate; the `setEventCallback` stream stays intact | `b2f843d` |
| 3 | `requestContentIds` static registry never cleaned → unbounded growth | `removeValue` on all 3 dismiss paths (inline, full-screen, display-error) | `b2f843d` |
| 4 | `catch (FormatException)` binds a var & catches everything (dart:663) | Explicit `catch (_)` + comment (broad catch is intentional — TypeError on non-String must be swallowed) | `b2f843d` |
| 5 | Same `DateTime.parse`/TypeError issue in `getBuiltInAttributes` (dart:680) | Same `catch (_)` across all 4 attribute readers | `b2f843d` |
| 6 | Android `getBuiltInAttribute` returns without `result` on missing `key` → hangs the Dart Future | `result.error("MISSING_PARAMETER", …)` (matches `setDebugMode`) | `acfc846` |

### Tests added
- **Dart** — `userAttribute returns a non-string value as-is` (pins the TypeError-swallow behaviour a narrow `on FormatException` would break). Full suite: **295 passed**.
- **Android** — `getBuiltInAttribute without key returns argument error instead of hanging`. Suite: **13 tests, 0 failures**.
- **iOS** (RunnerTests, same target as the existing compile-checks) — inline view is a `PLYEventDelegate`; `setEventDelegate`/`removeEventDelegate` API surface; `requestContentIds` registry clearable.
- **E2E** — new `inline_events_test.dart` asserts the global `listenToEvents` stream keeps flowing while an inline `PLYPresentationView` is mounted (FLT-W-12 regression guard), wired into both E2E runners (non-blocking until the inline-render path is proven reliable in CI). Note: `inline_paywall_test.dart` existed but was wired into neither runner — the inline path had no active E2E coverage before this.

### CI status
Dart ✅ / Android ✅. The **iOS example build failure is the pre-existing, out-of-scope `PLYDisplayMode`/`onCloseRequested`/`.screenId` symbol mismatch on the base branch** (documented in the PR description; fixed separately in `fix/ios-plytransition-type`) — no new error was introduced by these changes.

**Comment by kherembourg** [comment]
@greptileai review

**Review by macminisupport — APPROVED** [review]

**Comment by kherembourg** [comment]
### Rebased onto #132 — CI now fully green ✅

This PR was mistakenly based on `feat/sdk-v6-migration`, so its iOS build was red on the pre-existing `PLYDisplayMode`/`onClose`/`.id` v5 symbols. **Base changed to `fix/ios-plytransition-type` (#132)** and the branch rebased onto it — #132 carries the v5→v6 symbol renames (`PLYTransition`, `onCloseRequested`, `.screenId`), so the iOS build now compiles. The rebase auto-merged cleanly.

Additional CI fixes in this PR:
- **iOS unit test job was a false green** (`xcodebuild test … | xcpretty … || true` + a `iPhone 15,OS=latest` destination that no longer exists on `macos-latest`). Removed `|| true`; the RunnerTests host (the full Flutter example app) SIGSEGVs on the *headless* CI simulator, so the job now runs `build-for-testing` on a generic simulator destination — a real API-drift compile gate (these tests are compile-checks by design), verified locally with `** TEST BUILD SUCCEEDED **`.
- Added `NSBonjourServices` + `NSLocalNetworkUsageDescription` to the example `Info.plist` (Dart VM service on iOS 14+).
- `example/android/app/build.gradle` pinned to `minSdkVersion 23` (CI's `flutter.minSdkVersion` is 21 < the SDK's required 23).

**CI (29597156430): all 9 jobs green, including the `CI Success` gate** — Analyze & Format, Unit Tests (Dart), Android Unit Tests, iOS Unit Tests, Build iOS, Build Android, Validate Podspec, Version Consistency.

**Comment by kherembourg** [comment]
### iOS unit tests now execute for real on CI ✅

Converted `RunnerTests` to a **hostless logic-test bundle** so the tests actually run instead of just compiling:
- `pbxproj`: removed `TEST_HOST` / `BUNDLE_LOADER` / `TestTargetID` — the XCTest host was the full Flutter example app, which SIGSEGVs on launch on headless CI simulators.
- `Podfile`: RunnerTests `inherit! :complete` (was `:search_paths`) so the test bundle links/embeds `Purchasely` + `purchasely_flutter` itself (no host to load them from); `@testable import` still works.
- `ci.yml`: back to `xcodebuild test -only-testing:RunnerTests` on a dynamically-resolved booted simulator.

**CI (29726222239): iOS Unit Tests → `Executed 13 tests, with 0 failures`.** All 9 jobs green including the `CI Success` gate. No more `|| true` false-green.
